homeassistant/components/intellifire/__init__.py

```diff
@@ -31,82 +42,165 @@
+        cloud_interface.user_data.get_data_for_serial(serial) if valid_serial else None
+    )
+    if not new_data:
+        new_data = cloud_interface.user_data.get_data_for_ip(new[CONF_IP_ADDRESS])
+
+    if not new_data:
+        raise ConfigEntryAuthFailed
+
+    # Find the correct fireplace
+    if new_data is not None:
+        new[CONF_API_KEY] = new_data.api_key
+        new[CONF_WEB_CLIENT_ID] = new_data.web_client_id
+        new[CONF_AUTH_COOKIE] = new_data.auth_cookie
+
+        new[CONF_IP_ADDRESS] = new_data.ip_address
+        new[CONF_SERIAL] = new_data.serial
+
+        config_entry.version = 1
+        hass.config_entries.async_update_entry(
+            config_entry,
+            data=new,
+            options={CONF_READ_MODE: "local", CONF_CONTROL_MODE: "local"},
+            unique_id=serial,
+        )
+        LOGGER.debug("Pseudo Migration %s successful", config_entry.version)
+
+    return config_entry
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up IntelliFire from a config entry."""
     LOGGER.debug("Setting up config entry: %s", entry.unique_id)
 
     if CONF_USERNAME not in entry.data:
-        LOGGER.debug("Old config entry format detected: %s", entry.unique_id)
+        LOGGER.debug("Super Old config entry format detected: %s", entry.unique_id)
         raise ConfigEntryAuthFailed
 
-    ift_control = IntellifireControlAsync(
-        fireplace_ip=entry.data[CONF_HOST],
+    if CONF_IP_ADDRESS not in entry.data:
+        LOGGER.debug("Old config entry format detected: %s", entry.unique_id)
+        entry = await _async_pseudo_migrate_entry(hass, entry)
+
+    # Build a common data object to pass to the coordinator
+    fireplace: UnifiedFireplace = await UnifiedFireplace.build_fireplace_from_common(
+        _construct_common_data(entry)
     )
-    try:
-        await ift_control.login(
-            username=entry.data[CONF_USERNAME],
-            password=entry.data[CONF_PASSWORD],
-        )
-    except (ConnectionError, ClientConnectionError) as err:
-        raise ConfigEntryNotReady from err
-    except LoginException as err:
-        raise ConfigEntryAuthFailed(err) from err
 
-    finally:
-        await ift_control.close()
+    # Validate connectivity
+    local_connect, cloud_connect = await fireplace.async_validate_connectivity(
+        timeout=30
+    )
 
-    # Extract API Key and User_ID from ift_control
-    # Eventually this will migrate to using IntellifireAPICloud
+    LOGGER.info(
+        f"IntelliFire Connectivity: Local[{local_connect}]  Cloud[{cloud_connect}]"
+    )
 
-    if CONF_USER_ID not in entry.data or CONF_API_KEY not in entry.data:
-        LOGGER.info(
-            "Updating intellifire config entry for %s with api information",
-            entry.unique_id,
+    # If neither Local nor Cloud works - raise an Authentication issue
+    if (local_connect, cloud_connect) == (False, False):
+        raise ConfigEntryAuthFailed(
```

Comment:
Fireplace couldn't be hit by local IP address and we couldn't log into the cloud to access it so I think we need to reauthenticate again - is this correct?

Reply:
>Raise the ConfigEntryAuthFailed exception, and Home Assistant will automatically put the config entry in a failure state and start a reauth flow.

So yes I think this is correct in fact :)